crates/tags/src/lib.rs

```diff
@@ -95,10 +98,131 @@ where
+                                eprintln!("Failed to get tag metadata: {}", e);
+                                (StatusCode::INTERNAL_SERVER_ERROR, "Storage backend failure")
+                            }
+                        })
+                        .map(AsyncReadExt::compat)
+                        .map(serde_json::from_reader)?
+                        .map_err(|e| {
+                            eprintln!("Failed to read tag metadata: {}", e);
+                            (StatusCode::INTERNAL_SERVER_ERROR, "Storage backend failure")
+                        })?;
+
+                    let mut br = s.get(body_key(&tag.0)).await.map_err(|e| match e {
+                        GetError::NotFound => (StatusCode::NOT_FOUND, "Tag does not exist"),
+                        GetError::Internal(e) => {
+                            eprintln!("Failed to get tag: {}", e);
+                            (StatusCode::INTERNAL_SERVER_ERROR, "Storage backend failure")
+                        }
+                    })?;
+                    // TODO: Stream body
+                    // probably there should be a way to write body within the closure
+                    let mut body = vec![];
+                    br.read_to_end(&mut body).await.map_err(|e| {
+                        eprintln!("Failed to read tag contents: {}", e);
+                        (StatusCode::INTERNAL_SERVER_ERROR, "Storage backend failure")
+                    })?;
+                    Ok::<_, (_, _)>((meta, body))
+                }
+            }),
+        )
+        .route(
+            "/:tag",
+            put({
+                let s = s.clone();
+                move |Path(tag): Path<Name>, body: BodyStream, _: Tag, meta: Meta| async move {
+                    // TODO: Validate body as it's being read
+                    let mut s = s.write().await;
+
+                    let mw = s.create(meta_key(&tag.0)).await.map_err(|e| match e {
+                        CreateError::Occupied => (StatusCode::CONFLICT, "Tag already exists"),
+                        CreateError::Internal(e) => {
+                            eprintln!("Failed to create tag metadata: {}", e);
+                            (StatusCode::INTERNAL_SERVER_ERROR, "Storage backend failure")
+                        }
+                    })?;
+                    serde_json::to_writer(mw.compat_write(), &meta).map_err(|e| {
+                        eprintln!("Failed to read tag metadata: {}", e);
+                        (StatusCode::INTERNAL_SERVER_ERROR, "Storage backend failure")
+                    })?;
+
+                    let mut bw = s.create(body_key(&tag.0)).await.map_err(|e| match e {
+                        CreateError::Occupied => (StatusCode::CONFLICT, "Tag already exists"),
+                        CreateError::Internal(e) => {
+                            eprintln!("Failed to create tag: {}", e);
+                            (StatusCode::INTERNAL_SERVER_ERROR, "Storage backend failure")
```

Comment:
We need some common functionality for error conversion, but could be done as a follow-up for speed of iterations